httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/gss/GssSchemeBase.java

```diff
@@ -0,0 +1,366 @@
+        this.dnsResolver = dnsResolver != null ? dnsResolver : SystemDefaultDnsResolver.INSTANCE;
+        this.mutualAuth = config.isRequestMutualAuth();
+        this.state = State.UNINITIATED;
+    }
+
+    GssSchemeBase(final GssConfig config) {
+        this(config, SystemDefaultDnsResolver.INSTANCE);
+    }
+
+    GssSchemeBase() {
+        this(GssConfig.DEFAULT, SystemDefaultDnsResolver.INSTANCE);
+    }
+
+    @Override
+    public String getRealm() {
+        return null;
+    }
+
+    // Required by AuthScheme for backwards compatibility
+    @Override
+    public void processChallenge(final AuthChallenge authChallenge,
+            final HttpContext context ) {
+        // If this gets called, then AuthScheme was changed in an incompatible way
+        throw new UnsupportedOperationException();
+    }
+
+    // The AuthScheme API maps awkwardly to GSSAPI, where proccessChallange and generateAuthResponse
+    // map to the same single method call. Hence the generated token is only stored in this method.
+    @Override
+    public void processChallenge(
+            final HttpHost host,
+            final boolean challenged,
+            final AuthChallenge authChallenge,
+            final HttpContext context
+            ) throws AuthenticationException {
+
+        if (challengesLeft-- <= 0 ) {
+            if (LOG.isDebugEnabled()) {
+                final HttpClientContext clientContext = HttpClientContext.cast(context);
+                final String exchangeId = clientContext.getExchangeId();
+                LOG.debug("{} GSS error: too many challenges received. Infinite loop ?", exchangeId);
+            }
+            // TODO: Should we throw an exception ? There is a test for this behaviour.
+            state = State.FAILED;
+            return;
+        }
+
+        final byte[] challengeToken = Base64.decodeBase64(authChallenge == null ? null : authChallenge.getValue());
+
+        final String gssHostname;
+        String hostname = host.getHostName();
+        if (config.isUseCanonicalHostname()) {
+            try {
+                 hostname = dnsResolver.resolveCanonicalHostname(host.getHostName());
+            } catch (final UnknownHostException ignore) {
+            }
+        }
+        if (config.isAddPort()) {
+            gssHostname = hostname + ":" + host.getPort();
+        } else {
+            gssHostname = hostname;
+        }
+
+        if (LOG.isDebugEnabled()) {
+            final HttpClientContext clientContext = HttpClientContext.cast(context);
+            final String exchangeId = clientContext.getExchangeId();
+            LOG.debug("{} GSS init {}", exchangeId, gssHostname);
+        }
+        try {
+            setGssCredential(HttpClientContext.cast(context).getCredentialsProvider(), host, context);
+            queuedToken = generateToken(challengeToken, KERBEROS_SCHEME, gssHostname);
+            switch (state) {
+            case UNINITIATED:
+                if (challenge != NO_TOKEN) {
```

Comment:
String and `""`? Why not work with `null`?

Reply:
That's completely buggy, we never set that variable.
I'm going to fix that.